server/repository/src/db_diesel/temperature_chart/temperature_chart.rs

```diff
@@ -0,0 +1,355 @@
+
+impl<'a> TemperatureChartRepository<'a> {
+    pub fn new(connection: &'a StorageConnection) -> Self {
+        TemperatureChartRepository { connection }
+    }
+
+    pub fn query(
+        &self,
+        from_datetime: NaiveDateTime,
+        to_datetime: NaiveDateTime,
+        number_of_data_points: i32,
+        temperature_log_filter: Option<TemperatureLogFilter>,
+    ) -> Result<(Vec<TemperatureChartRow>, Vec<Interval>), RepositoryError> {
+        let intervals = calculate_intervals(from_datetime, to_datetime, number_of_data_points);
+        let mut query = TemperatureChart {
+            intervals: intervals.clone(),
+        }
+        .into_boxed::<DBType>();
+
+        if temperature_log_filter.is_some() {
+            let temperature_log_ids =
+                TemperatureLogRepository::create_filtered_query(temperature_log_filter)
+                    .select(temperature_log_dsl::id);
+            query = query.filter(TemperatureLogId.eq_any(temperature_log_ids));
+        };
+
+        let query = query
+            .select((FromDatetime, ToDatetime, AverageTemperature, SensorId))
+            .group_by((FromDatetime, ToDatetime, SensorId));
+
+        // First by sensor then by datetime (so should be sorted by sensor and then by datetime)
+        let query = query.order_by((SensorId.asc(), FromDatetime.asc()));
+
+        // Debug diesel query
+        // println!("{}", diesel::debug_query::<DBType, _>(&query).to_string());
+
+        let chart_data = query
+            .load::<QueryResult>(&self.connection.connection)?
+            .into_iter()
+            .map(TemperatureChartRow::from)
+            .collect();
+
+        Ok((chart_data, intervals))
+    }
+}
+
+impl From<QueryResult> for TemperatureChartRow {
+    fn from((from_datetime, to_datetime, average_temperature, sensor_id): QueryResult) -> Self {
+        Self {
+            from_datetime,
+            to_datetime,
+            average_temperature,
+            sensor_id,
+        }
+    }
+}
+
+fn calculate_intervals(
+    from_datetime: NaiveDateTime,
+    to_datetime: NaiveDateTime,
+    number_of_data_points: i32,
+) -> Vec<Interval> {
+    let interval = (to_datetime - from_datetime) / number_of_data_points;
+    (0..number_of_data_points)
+        .map(|point| Interval {
+            from_datetime: datetime_with_offset(&from_datetime, interval * point),
+            to_datetime: datetime_with_offset(&from_datetime, interval * (point + 1)),
+        })
+        .collect()
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::{
+        location::LocationFilter,
+        mock::{MockData, MockDataInserts},
+        test_db::setup_all_with_data,
+        EqualFilter, LocationRow, NameRow, SensorFilter, SensorRow, StoreRow,
+        TemperatureChartRepository, TemperatureChartRow, TemperatureLogRow,
+    };
+    use util::create_datetime;
+    #[test]
+    fn test_calculate_intervals() {
+        assert_eq!(
```

Comment:
> I'm a little concerned what we should expect if even intervals require milliseconds, but we have seconds resolution in the db?

Great point, I didn't even think about this, was trusting the diesel/chrono implementation ? 

Added a test: 195f1d77c3ccb2cb26691897f3474be35b649ba2

Although I think your question may also be related to how we group temperatures for those intervals ? I can't quite think of a problem with grouping, please let me know if you can think of a use case.

> Ideally with some english explanations, something like this?

Sorry I thought the test was pretty straight forward, added small comment195f1d77c3ccb2cb26691897f3474be35b649ba2

> but I'd like to see more calculate interval tests here.

Do you have a suggestion of another test ? Happy to add a new one if you think it will cover more 

(the test I have covers midnight transition, I thought it was sufficient)

Reply:
> Sorry I thought the test was pretty straight forward, added small comment https://github.com/msupply-foundation/open-msupply/commit/211c998126bbf931273cf8412887a9d6e9b155bd

You're right, it's pretty straighforward and readable.
I was assuming we might have more than one test case, and thought it would be good to add more of a `Context` comment for each one. I'll add a suggestion to try to explain better.

It's not a major issue at all, I'm trying to improve my test comments too, so was probably in my mind!